### ISSUE TITLE
Collaboration Mode A - Full Featured Collaboration

### ISSUE THREAD

**Issue by flukeout** [mozilla/thimble.mozilla.org#1851]
This collaboration feature set closely mimics what we're used to already from services like Google Docs. Projects can have multiple contributors that are able to edit the project at any time. Here's a detailed feature summary:

As the project owner, I can set the following types of collaboration permissions for your project:
    
* This project can only be edited by me (default)
* This project can be edited be me and collaborators that I explicitly specify by their username
* This project can be edited by anyone who has the link

I can change these settings at any time.

Collaborators can, at any time can...

* Edit the contents of the files within my project
* Edit the contents of the file tree within my project

Collaborators can't...
* Delete my project
* Publish or unpublish my project

**Real-time Collaboration**
When one of my collaborators is editing the Project at the same time that I am...

* I can see their username and avatar
* If they're editing the same file as I am
* I can see the changes to the file in real-time
* I can see the position of their cursor, and selection in real time

If they're editing a different file
* I can see which file they are editing

**Comment by Pomax** [comment]
should I also be able to "block" collaborators if it turns out they are more into destroying my work, and unroll the changes they made? (e.g. bad egg collaborator, pranksters, etc)

**Comment by AnshulMalik** [comment]
I'd like to contribute when the design is finalized :)

**Comment by flukeout** [comment]
That's great @AnshulMalik - you can also look through all of the issues labeled Collaboration to get more familiar with the project https://github.com/mozilla/thimble.mozilla.org/labels/Collaboration

**Comment by AnshulMalik** [comment]
Hey, I was just going through GSoC projects from mozilla, and I am glad that thimble is also there.

So that's a plus point for me, since I can get paid as well :P

**Comment by thekyriarchy** [comment]
Does this mean that collaboration is no longer possible as it was on the old Webmaker's Thimble app? The TogetherJS website says that Thimble is currently using it to enable collaboration, but the screenshot is years out of date... https://github.com/mozilla/togetherjs/issues/1126

**Comment by ryanwarsaw** [comment]
@thekyriarchy You're correct, that screenshot is out-of-date and I don't think that page as a whole is updated as regularly as it might be needed. Collaborative editing has been in the works for some time, there was a lot of work done on it over the summer although I'm not sure what has happened since then.
